### DISCUSSION TITLE
Too many chunks

### DISCUSSION THREAD

**Discussion by noinhotnn** [seaweedfs/seaweedfs#2876] [Q&A]
Every day my system receives hundreds of millions of records. They are evenly distributed throughout the day (every second). I put it use append but metadata is huge and weed use huge memory to store it. I know that:
`When appending to a file, each append will create one chunk and added to the file metadata. If there are too many small appends, there could be too many chunks. So try to keep each append size reasonably big.`
My question:
How do i merger chunks at the end of day in to one to reduce metadata and memory.

**Comment by chrislusf** [comment]
Maybe buffer the changes before writing.

**Reply by noinhotnn** [reply]
In your way. First, need too much memory. Second, latency of data is important for customer.

**Reply by chrislusf** [reply]
Seems you want a message queue, which is in the dev plan but not started yet.

**Reply by kmlebedev** [reply]
@noinhotnn I don't quite understand what the problem is?
What client interface are you using?
s3 client allows you to configure the maximum size of chunks.
The filer also allows you to configure the maximum size of chunks for the entire cluster
```
weed filer --help
Example: weed filer -port=8888 -master=<ip:port>[,<ip:port>]*
Default Usage:
  -maxMB int
    	split files larger than the limit (default 4)
```

@chrislusf  Maybe add a setting for the maximum size of chunks for the collection?
example:
```
fs.configure --help
Usage of fs.configure:
  -maxMB int
    	split files larger than the limit (default 4)
```

And how can message queue help with this?

**Reply by noinhotnn** [reply]
@chrislusf i don't think message queue is good idea for my problem (i worked kafka, rabbitmq...)
Hi @kmlebedev. It not a problem for size of chunks. I used S3 custom to put append. If you push the file in one time, it only generates a block for management (16 Byte?). If you split the file into many parts, it will generate many blocks (spend more memory).

**Reply by kmlebedev** [reply]
@noinhotnn As I understand it, the problem is that on the client side you split the file into extremely small parts, and then you want to collect it into one large file?
What values are we talking about?
What is the size of one chunk?
What is the average number of chunks for one file?
What is the size of the resulting file?

**Reply by noinhotnn** [reply]
`As I understand it, the problem is that on the client side you split the file into extremely small parts, and then you want to collect it into one large file?`
Right, My device periodically sends data to the server. At the beginning of the day i create file to store all that data
`What values are we talking about?`
I'm not sure about seaweed architecture but I think blocks when i append to exists file, block is increase.
`What is the size of one chunk?`
About 1-2KB
`What is the average number of chunks for one file?`
Average, the device sends a signal about once every 20 seconds so it about ~4k3 chunks
`What is the size of the resulting file?`
Every day one file stores about 3-4MB

**Comment by chrislusf** [comment]
I added https://github.com/chrislusf/seaweedfs/commit/79b8e6a8c31d20b309f3ffe67fb37701054d6000 and a TODO

**Reply by kmlebedev** [reply]
@chrislusf It seems that you can check the size of chunks and merge if the size is much less than 5 mb by default
```
S3 requires a minimum chunk size of 5MB, and supports at most 10,000 chunks per multipart upload.
```
https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html

**Comment by tonyhb** [comment]
@chrislusf continual appending to an "in-progress" blob is something that I'm interested in for a variety of reasons (namely, read perf).

I like your current "merge when writing" todo, and this is my ideal.  Could I help with that?

Also, we could have a background process that "merges" appended chunks once they reach a specific threshold into a single consolidated chunk.

How it could work (naively, without reading any existing code):

1. On append, check the count of chunks
2. If greater than a threshold, mark for consolidation (in some other separate key / metadata / list)
3. Periodically run a consolidation cleanup, which:
4. Performs a snapshot read on the file's metadata to grab the currently existing and completed chunks
5. Consolidates the chunks into a single, new chunk.
6. Replaces the split chunks with the single chunk in the file's metadata.
7. Marks the previous split chunks as deleted for future GC.

Because this is a snapshot read, the file's metadata should still be modifiable with concurrent appends.  The consolidation logic has access to the chunk IDs it's replacing.

**Comment by chrislusf** [comment]
Currently it already merge many chunk ids into one chunk list and saves into a meta chunk.

**Reply by tonyhb** [reply]
Okay, nice!  This is now just for curiosity:  does this meta chunk still require N filer store lookups and blob reads?  I'm  whether background aggregation that merges eg ~1,000 individual chunks and "meta chunks" into a single chunk would make subsequent reads faster :)
